File: turms-client-swift/Sources/TurmsClient/Service/MessageService.swift
import Foundation
import PromiseKit

public class MessageService {
    /**
     * Format: "@{userId}"
     * Example: "@{123}", "I need to talk with @{123} and @{321}"
     */
    private static let DEFAULT_MENTIONED_USER_IDS_REGEX = try! NSRegularExpression(pattern: "@\\{(\\d+?)\\}", options: [])
    private static let DEFAULT_MENTIONED_USER_IDS_PARSER: (_ message: Message) -> [Int64] = {
        if $0.hasText {
            let text = $0.text
            let results = DEFAULT_MENTIONED_USER_IDS_REGEX.matches(in: text, range: NSRange(text.startIndex..., in: text))
            return results.map {
                Int64(String(text[Range($0.range(at: 1), in: text)!]))!
            }
        }
        return []
    }

    private weak var turmsClient: TurmsClient!
    private var mentionedUserIdsParser: ((Message) -> [Int64])?
    public var messageListeners: [(Message, MessageAddition) -> Void] = []

    init(_ turmsClient: TurmsClient) {
        self.turmsClient = turmsClient
        self.turmsClient.driver
            .addNotificationListener {
                if !self.messageListeners.isEmpty, $0.hasRelayedRequest, case let .createMessageRequest(request) = $0.relayedRequest.kind {
                    let message = MessageService.createMessage2Message($0.requesterID, request)
                    let addition = self.parseMessageAddition(message)
                    self.messageListeners.forEach { listener in listener(message, addition) }
                }
            }
    }

    func addMessageListener(_ listener: @escaping (Message, MessageAddition) -> Void) {
        messageListeners.append(listener)
    }

    public func sendMessage(
        isGroupMessage: Bool,
        targetId: Int64,
        deliveryDate: Date? = nil,
        text: String? = nil,
        records: [Data]? = nil,
        burnAfter: Int32? = nil,
        preMessageId: Int64? = nil
    ) -> Promise<Int64> {
        if Validator.areAllNil(text, records) {
            return Promise(error: TurmsBusinessError(TurmsStatusCode.illegalArgument, "text and records must not all be null"))
        }
        return turmsClient.driver
            .send {
                $0.createMessageRequest = .with {
                    if isGroupMessage {
                        $0.groupID = targetId
                    } else {
                        $0.recipientID = targetId
                    }
                    if let v = deliveryDate {
                        $0.deliveryDate = v.toMillis()
                    }
                    if let v = text {
                        $0.text = v
                    }
                    if let v = records {
                        $0.records = v
                    }
                    if let v = burnAfter {
                        $0.burnAfter = v
                    }
                    if let v = preMessageId {
                        $0.preMessageID = v
                    }
                }
            }
            .map {
                try $0.getFirstId()
            }
    }

    public func forwardMessage(
        messageId: Int64,
        isGroupMessage: Bool,
        targetId: Int64
    ) -> Promise<Int64> {
        return turmsClient.driver
            .send {
                $0.createMessageRequest = .with {
                    $0.messageID = messageId
                    if isGroupMessage {
                        $0.groupID = targetId
                    } else {
                        $0.recipientID = targetId
                    }
                }
            }
            .map {
                try $0.getFirstId()
            }
    }

    public func updateSentMessage(
        messageId: Int64,
        text: String? = nil,
        records: [Data]? = nil
    ) -> Promise<Void> {
        if Validator.areAllNil(text, records) {
            return Promise.value(())
        }
        return turmsClient.driver
            .send {
                $0.updateMessageRequest = .with {
                    $0.messageID = messageId
                    if let v = text {
                        $0.text = v
                    }
                    if let v = records {
                        $0.records = v
                    }
                }
            }
            .asVoid()
    }

    public func queryMessages(
        ids: [Int64]? = nil,
        areGroupMessages: Bool? = nil,
        areSystemMessages: Bool? = nil,
        fromId: Int64? = nil,
        deliveryDateAfter: Date? = nil,
        deliveryDateBefore: Date? = nil,
        size: Int32 = 50
    ) -> Promise<[Message]> {
        return turmsClient.driver
            .send {
                $0.queryMessagesRequest = .with {
                    if let v = ids {
                        $0.ids = v
                    }
                    if let v = areGroupMessages {
                        $0.areGroupMessages = v
                    }
                    if let v = areSystemMessages {
                        $0.areSystemMessages = v
                    }
                    if let v = fromId {
                        $0.fromID = v
                    }
                    if let v = deliveryDateAfter {
                        $0.deliveryDateAfter = v.toMillis()
                    }
                    if let v = deliveryDateBefore {
                        $0.deliveryDateBefore = v.toMillis()
                    }
                    $0.size = size
                    $0.withTotal = false
                }
            }
            .map {
                $0.data.messages.messages
            }
    }

    public func queryMessagesWithTotal(
        ids: [Int64]? = nil,
        areGroupMessages: Bool? = nil,
        areSystemMessages: Bool? = nil,
        fromId: Int64? = nil,
        deliveryDateAfter: Date? = nil,
        deliveryDateBefore: Date? = nil,
        size: Int32 = 1
    ) -> Promise<[MessagesWithTotal]> {
        return turmsClient.driver
            .send {
                $0.queryMessagesRequest = .with {
                    if let v = ids {
                        $0.ids = v
                    }
                    if let v = areGroupMessages {
                        $0.areGroupMessages = v
                    }
                    if let v = areSystemMessages {
                        $0.areSystemMessages = v
                    }
                    if let v = fromId {
                        $0.fromID = v
                    }
                    if let v = deliveryDateAfter {
                        $0.deliveryDateAfter = v.toMillis()
                    }
                    if let v = deliveryDateBefore {
                        $0.deliveryDateBefore = v.toMillis()
                    }
                    $0.size = size
                    $0.withTotal = true
                }
            }
            .map {
                $0.data.messagesWithTotalList.messagesWithTotalList
            }
    }

    public func recallMessage(messageId: Int64, recallDate: Date = Date()) -> Promise<Void> {
        return turmsClient.driver
            .send {
                $0.updateMessageRequest = .with {
                    $0.messageID = messageId
                    $0.recallDate = recallDate.toMillis()
                }
            }
            .asVoid()
    }

    public func isMentionEnabled() -> Bool {
        return mentionedUserIdsParser != nil
    }

    public func enableMention(mentionedUserIdsParser: ((Message) -> [Int64])?) {
        if mentionedUserIdsParser != nil {
            self.mentionedUserIdsParser = mentionedUserIdsParser
        } else if self.mentionedUserIdsParser == nil {
            self.mentionedUserIdsParser = MessageService.DEFAULT_MENTIONED_USER_IDS_PARSER
        }
    }

    public static func generateLocationRecord(latitude: Float, longitude: Float, locationName: String? = nil, address: String? = nil) -> Data {
        return try! UserLocation.with {
            $0.latitude = latitude
            $0.longitude = longitude
            if locationName != nil {
                $0.name = locationName!
            }
            if address != nil {
                $0.address = address!
            }
        }.serializedData()
    }

    public static func generateAudioRecordByDescription(url: String, duration: Int32? = nil, format: String? = nil, size: Int32? = nil) -> Data {
        return try! AudioFile.with {
            $0.description_p.url = url
            if duration != nil {
                $0.description_p.duration = duration!
            }
            if format != nil {
                $0.description_p.format = format!
            }
            if size != nil {
                $0.description_p.size = size!
            }
        }.serializedData()
    }

    public static func generateAudioRecordByData(_ data: Data) -> Data {
        return try! AudioFile.with {
            $0.data = data
        }.serializedData()
    }

    public static func generateVideoRecordByDescription(url: String, duration: Int32? = nil, format: String? = nil, size: Int32? = nil) -> Data {
        return try! VideoFile.with {
            $0.description_p.url = url
            if duration != nil {
                $0.description_p.duration = duration!
            }
            if format != nil {
                $0.description_p.format = format!
            }
            if size != nil {
                $0.description_p.size = size!
            }
        }.serializedData()
    }

    public static func generateVideoRecordByData(_ data: Data) -> Data {
        return try! VideoFile.with {
            $0.data = data
        }.serializedData()
    }

    public static func generateImageRecordByData(_ data: Data) -> Data {
        return try! ImageFile.with {
            $0.data = data
        }.serializedData()
    }

    public static func generateImageRecordByDescription(url: String, fileSize: Int32? = nil, imageSize: Int32? = nil, original: Bool? = nil) -> Data {
        return try! ImageFile.with {
            $0.description_p.url = url
            if fileSize != nil {
                $0.description_p.fileSize = fileSize!
            }
            if imageSize != nil {
                $0.description_p.imageSize = imageSize!
            }
            if original != nil {
                $0.description_p.original = original!
            }
        }.serializedData()
    }

    public static func generateFileRecordByDate(_ data: Data) -> Data {
        return try! File.with {
            $0.data = data
        }.serializedData()
    }

    public static func generateFileRecordByDescription(url: String, format: String? = nil, size: Int32? = nil) -> Data {
        return try! File.with {
            $0.description_p.url = url
            if format != nil {
                $0.description_p.format = format!
            }
            if size != nil {
                $0.description_p.size = size!
            }
        }.serializedData()
    }

    private func parseMessageAddition(_ message: Message) -> MessageAddition {
        let mentionedUserIds = mentionedUserIdsParser?(message) ?? []
        let userId = turmsClient.userService.userInfo?.userId
        let isMentioned = userId != nil ? mentionedUserIds.contains(userId!) : false
        let records = message.records
        var systemMessageType: BuiltinSystemMessageType?
        if message.isSystemMessage, !records.isEmpty {
            let data = records[0]
            if !data.isEmpty {
                systemMessageType = BuiltinSystemMessageType(rawValue: Int(data[0]))
            }
        }
        var recalledMessageIds: [Int64] = []
        if systemMessageType == BuiltinSystemMessageType.recallMessage {
            let size = records.count
            for i in 1 ... (size - 1) {
                let id = records[i].withUnsafeBytes {
                    $0.load(as: Int64.self)
                }
                recalledMessageIds.append(id)
            }
        }
        return MessageAddition(isMentioned: isMentioned, mentionedUserIds: mentionedUserIds, recalledMessageIds: recalledMessageIds)
    }

    private static func createMessage2Message(_ requesterId: Int64, _ request: CreateMessageRequest) -> Message {
        return Message.with {
            if request.hasMessageID {
                $0.id = request.messageID
            }
            $0.isSystemMessage = request.isSystemMessage
            $0.deliveryDate = request.deliveryDate
            if request.hasText {
                $0.text = request.text
            }
            if request.records.count > 0 {
                $0.records = request.records
            }
            $0.senderID = requesterId
            if request.hasGroupID {
                $0.groupID = request.groupID
            }
            if request.hasRecipientID {
                $0.recipientID = request.recipientID
            }
        }
    }
}

File: turms-client-dart/lib/src/service/message_service.dart
import 'dart:typed_data';

import 'package:fixnum/fixnum.dart' show Int64;

import '../../turms_client.dart';
import '../extension/date_time_extensions.dart';
import '../extension/iterable_extensions.dart';
import '../extension/notification_extensions.dart';
import '../model/builtin_system_message_type.dart';
import '../model/message_addition.dart';
import '../model/model/file/audio_file.pb.dart';
import '../model/model/file/file.pb.dart';
import '../model/model/file/image_file.pb.dart';
import '../model/model/file/video_file.pb.dart';
import '../model/model/message/message.pb.dart';
import '../model/model/message/messages_with_total.pb.dart';
import '../model/model/user/user_location.pb.dart';
import '../model/request/message/create_message_request.pb.dart';
import '../model/request/message/query_messages_request.pb.dart';
import '../model/request/message/update_message_request.pb.dart';
import '../model/turms_business_exception.dart';
import '../model/turms_status_code.dart';

typedef MentionedUserIdsParser = Set<Int64> Function(Message message);
typedef MessageListener = void Function(
    Message message, MessageAddition addition);

class MessageService {
  /// Format: "@{userId}"
  /// Example: "@{123}", "I need to talk with @{123} and @{321}"
  static final RegExp _defaultMentionedUserIdsParserRegex =
      RegExp(r'@{(\d+?)}', multiLine: true);

  Set<Int64> _defaultMentionedUserIdsParser(Message message) {
    if (message.hasText()) {
      final userIds = <Int64>{};
      for (final matches
          in _defaultMentionedUserIdsParserRegex.allMatches(message.text)) {
        userIds.add(Int64.parseInt(matches.group(1)!));
      }
      return userIds;
    }
    return {};
  }

  final TurmsClient _turmsClient;
  MentionedUserIdsParser? _mentionedUserIdsParser;
  final List<MessageListener> _messageListeners = [];

  MessageService(this._turmsClient) {
    _turmsClient.driver.addNotificationListener((notification) {
      if (_messageListeners.isNotEmpty &&
          notification.hasRelayedRequest() &&
          notification.relayedRequest.hasCreateMessageRequest()) {
        final request = notification.relayedRequest.createMessageRequest;
        final message =
            _createMessageRequest2Message(notification.requesterId, request);
        final addition = _parseMessageAddition(message);
        for (final listener in _messageListeners) {
          listener.call(message, addition);
        }
      }
    });
  }

  void addMessageListener(MessageListener listener) =>
      _messageListeners.add(listener);

  void removeMessageListener(MessageListener listener) =>
      _messageListeners.remove(listener);

  Future<Int64> sendMessage(bool isGroupMessage, Int64 targetId,
      {DateTime? deliveryDate,
      String? text,
      List<Uint8List>? records,
      int? burnAfter,
      Int64? preMessageId}) async {
    if (text == null && (records?.isEmpty ?? true)) {
      throw TurmsBusinessException(TurmsStatusCode.illegalArgument,
          'text and records must not all be null');
    }
    deliveryDate ??= DateTime.now();
    final n = await _turmsClient.driver.send(CreateMessageRequest(
        groupId: isGroupMessage ? targetId : null,
        recipientId: !isGroupMessage ? targetId : null,
        deliveryDate: deliveryDate.toInt64(),
        text: text,
        records: records,
        burnAfter: burnAfter,
        preMessageId: preMessageId));
    return n.getFirstIdOrThrow();
  }

  Future<Int64> forwardMessage(
      Int64 messageId, bool isGroupMessage, Int64 targetId) async {
    final n = await _turmsClient.driver.send(CreateMessageRequest(
        messageId: messageId,
        groupId: isGroupMessage ? targetId : null,
        recipientId: !isGroupMessage ? targetId : null));
    return n.getFirstIdOrThrow();
  }

  Future<void> updateSentMessage(Int64 messageId,
      {String? text, List<Uint8List>? records}) async {
    if ([text, records].areAllNull) {
      return;
    }
    await _turmsClient.driver.send(UpdateMessageRequest(
        messageId: messageId, text: text, records: records));
  }

  Future<List<Message>> queryMessages(
      {Set<Int64>? ids,
      bool? areGroupMessages,
      bool? areSystemMessages,
      Int64? fromId,
      DateTime? deliveryDateAfter,
      DateTime? deliveryDateBefore,
      int size = 50}) async {
    final n = await _turmsClient.driver.send(QueryMessagesRequest(
        ids: ids,
        areGroupMessages: areGroupMessages,
        areSystemMessages: areSystemMessages,
        fromId: fromId,
        deliveryDateAfter: deliveryDateAfter?.toInt64(),
        deliveryDateBefore: deliveryDateBefore?.toInt64(),
        size: size,
        withTotal: false));
    return n.data.messages.messages;
  }

  Future<List<MessagesWithTotal>> queryMessagesWithTotal(
      {Set<Int64>? ids,
      bool? areGroupMessages,
      bool? areSystemMessages,
      Int64? fromId,
      DateTime? deliveryDateAfter,
      DateTime? deliveryDateBefore,
      int size = 1}) async {
    final n = await _turmsClient.driver.send(QueryMessagesRequest(
        ids: ids,
        areGroupMessages: areGroupMessages,
        areSystemMessages: areSystemMessages,
        fromId: fromId,
        deliveryDateAfter: deliveryDateAfter?.toInt64(),
        deliveryDateBefore: deliveryDateBefore?.toInt64(),
        size: size,
        withTotal: true));
    return n.data.messagesWithTotalList.messagesWithTotalList;
  }

  Future<void> recallMessage(Int64 messageId, {DateTime? recallDate}) =>
      _turmsClient.driver.send(UpdateMessageRequest(
          messageId: messageId,
          recallDate:
              Int64((recallDate ?? DateTime.now()).millisecondsSinceEpoch)));

  bool isMentionEnabled() => _mentionedUserIdsParser != null;

  void enableMention({MentionedUserIdsParser? mentionedUserIdsParser}) {
    _mentionedUserIdsParser = mentionedUserIdsParser ??
        _mentionedUserIdsParser ??
        _defaultMentionedUserIdsParser;
  }

  static Uint8List generateLocationRecord(double latitude, double longitude,
          {String? locationName, String? address}) =>
      UserLocation(
              latitude: latitude,
              longitude: longitude,
              name: locationName,
              address: address)
          .writeToBuffer();

  static Uint8List generateAudioRecordByDescription(String url,
          {int? duration, String? format, int? size}) =>
      AudioFile(
          description: AudioFile_Description(
        url: url,
        duration: duration,
        format: format,
        size: size,
      )).writeToBuffer();

  static Uint8List generateAudioRecordByData(Uint8List data) =>
      AudioFile(data: data).writeToBuffer();

  static Uint8List generateVideoRecordByDescription(String url,
          {int? duration, String? format, int? size}) =>
      VideoFile(
          description: VideoFile_Description(
        url: url,
        duration: duration,
        format: format,
        size: size,
      )).writeToBuffer();

  static Uint8List generateVideoRecordByData(Uint8List data) =>
      VideoFile(data: data).writeToBuffer();

  static Uint8List generateImageRecordByData(Uint8List data) =>
      ImageFile(data: data).writeToBuffer();

  static Uint8List generateImageRecordByDescription(String url,
          {int? fileSize, int? imageSize, bool? original}) =>
      ImageFile(
              description: ImageFile_Description(
                  url: url,
                  fileSize: fileSize,
                  imageSize: imageSize,
                  original: original))
          .writeToBuffer();

  static Uint8List generateFileRecordByDate(Uint8List data) =>
      File(data: data).writeToBuffer();

  static Uint8List generateFileRecordByDescription(String url,
          {String? format, int? size}) =>
      File(description: File_Description(url: url, format: format, size: size))
          .writeToBuffer();

  MessageAddition _parseMessageAddition(Message message) {
    final mentionedUserIds = _mentionedUserIdsParser?.call(message) ?? {};
    final isMentioned =
        mentionedUserIds.contains(_turmsClient.userService.userInfo?.userId);
    final records = message.records;
    var systemMessageType = BuiltinSystemMessageType.normal;
    final recalledMessageIds = <Int64>{};
    if (message.isSystemMessage && records.isNotEmpty) {
      final bytes = records[0];
      if (bytes.isNotEmpty) {
        systemMessageType = BuiltinSystemMessageType.values
            .firstWhere((element) => element.index == bytes[0]);
        if (systemMessageType == BuiltinSystemMessageType.recallMessage) {
          final size = records.length;
          for (var i = 1; i < size; i++) {
            // Note that it will be truncated into the range of
            // -2^53 and 2^53 in JavaScript.
            final id = ByteData.view(Uint8List.fromList(records[i]).buffer)
                .getInt64(0);
            recalledMessageIds.add(Int64(id));
          }
        }
      }
    }
    return MessageAddition(isMentioned, mentionedUserIds, recalledMessageIds);
  }

  Message _createMessageRequest2Message(
          Int64 requesterId, CreateMessageRequest request) =>
      Message(
          id: request.messageId,
          isSystemMessage: request.isSystemMessage,
          deliveryDate: request.deliveryDate,
          text: request.text,
          records: request.records,
          senderId: requesterId,
          groupId: request.groupId,
          recipientId: request.recipientId);
}
